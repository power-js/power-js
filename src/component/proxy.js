import { isArray, isFunction } from '../utils/is';

const ARRAY_MODIFIERS = ['push', 'pop', 'shift', 'unshift', 'splice'];

/**
 * Wraps our prop/state objects to we can control writes and/or detect prop modifications specifically with Arrays
 * @param   {Class}   Component
 * @param   {Object}  source
 * @param   {Boolean} isStateObject
 * @return  {Proxy}
 */
export const proxy = (component, source) => {
  const handler = {
    get(target, prop) {
      const value = target[prop];

      const arrayHandler = {
        get(target, prop) {
          const value = target[prop];
          if (isFunction(value)) {
            if (isArray(target)) {
              // allow operation through
              return function() {
                const array = Array.prototype[prop].apply(target, arguments);

                if (ARRAY_MODIFIERS.includes(prop)) {
                  if (component.shouldComponentUpdate(component.props, component.state)) {
                    component.rerender();
                  }
                }

                return array;
              };
            }
          }

          return value;
        }
      };

      if (isArray(value)) {
        return new Proxy(value, arrayHandler);
      }

      return value;
    },
    set(target, prop, value) {
      const currentValue = target[prop];

      if (!component.shouldComponentUpdate(target, component.state)) {
        return true;
      }

      if (currentValue !== value) {
        // set value
        target[prop] = value;
        // update component
        component.rerender();
      }
      // return true to indicate that assignment succeeded
      return true;
    }
  };

  return new Proxy(source, handler);
};
