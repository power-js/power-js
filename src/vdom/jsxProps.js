/**
 * JSX Special Properties
 */
export const jsxProps = {
  htmlFor: 'for',
  className: 'class'
};
