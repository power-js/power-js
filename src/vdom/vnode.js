import { DATA_NODE_ATTRIBUTE } from '../constants';

/**
 * VNode Counter
 * @type {Number}
 */
let _counter = 0;

/**
 * Creates a Virtual Node
 * @public
 * @param   {String}  tag
 * @param   {Object}  props
 * @param   {Array}   children
 * @returns {Object}
 */
export function VNode(tagName, props, children) {
  this.tagName = tagName || 'div';
  this.children = children || [];
  this.props = props || {};

  _counter += 1;

  this.props[DATA_NODE_ATTRIBUTE] = _counter;

  return this;
}
