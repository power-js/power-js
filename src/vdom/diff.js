import { DATA_NODE_ATTRIBUTE } from '../constants';
import { isEvent, isString, isVNode, isEqual } from '../utils/is';
import { createElement } from '../dom/createElement';
import { updateElementStyles } from '../dom/updateElementStyles';
import { removeNode } from '../utils/dom/removeNode';
import { jsxProps } from './jsxProps';

/**
 * working on difference between 2 vnodes
 * @private
 * @param {Object}  oldVNode
 * @param {Object}  newVNode
 * @param {Class}   Component
 */
export const diff = (oldVNode, newVNode, Component) => {
  // get the element id
  const powerId = oldVNode.props[DATA_NODE_ATTRIBUTE];

  // check if newVNode props is null
  if (newVNode.props === null) {
    newVNode.props = {};
  }

  // merge the node id
  newVNode.props[DATA_NODE_ATTRIBUTE] = oldVNode.props[DATA_NODE_ATTRIBUTE];

  // get the dom element to the vnode
  const element = Component.node.querySelector(`[${DATA_NODE_ATTRIBUTE}="${powerId}"]`);

  // compare the tag
  if (oldVNode.tagName !== newVNode.tagName) {
    console.log('tagName changed');
  }

  // compare props
  propsDiff(oldVNode.props, newVNode.props, element);

  // compare children
  childrenDiff(oldVNode.children, newVNode.children, element, Component);
};

/**
 * checks out the difference between 2 objects
 * and merges it into the component element
 * @private
 * @param {Object}      oldObj
 * @param {Object}      newObj
 * @param {HTMLElement} element
 */
export const propsDiff = (oldObj, newObj, element) => {
  // prevent unneeded iterations
  if (isEqual(oldObj, newObj)) {
    return;
  }

  if (!isEqual(oldObj.style, newObj.style)) {
    // update styling
    updateElementStyles(element, newObj.style, oldObj.style);
  }

  for (const key in oldObj) {
    if (!newObj[key]) {
      // removing attribute from element
      element.removeAttribute(jsxProps[key] || key);
    }
  }

  for (const key in newObj) {
    if (key !== 'style' && !isEvent(key)) {
      // check if there a new key
      if (!oldObj[key] || newObj[key] !== oldObj[key]) {
        // add attribute to element
        element.setAttribute(jsxProps[key] || key, newObj[key]);
      }
    }
  }
};

/**
 * checks out the difference between 2 Arrays
 * and merges it into the component element
 * @private
 * @param {Array}       oldChildren
 * @param {Array}       newChildren
 * @param {HTMLElement} element
 * @param {Class}       Component
 */
export const childrenDiff = (oldChildren, newChildren, element, Component) => {
  for (let i = 0, k = newChildren.length; i < k; i++) {
    const child = newChildren[i];

    if (oldChildren[i] === undefined && isVNode(child)) {
      if (!child.props) {
        child.props = {};
      }

      const newElement = createElement(child, Component);

      element.appendChild(newElement);

      continue;
    }

    if (isString(child) && child !== oldChildren[i]) {
      const childText = document.createTextNode(child);
      element.replaceChild(childText, element.childNodes[i]);

      continue;
    }

    if (isVNode(child)) {
      diff(oldChildren[i], child, Component);

      continue;
    }

    if (child.pop && oldChildren[i] && oldChildren[i].pop) {
      childrenDiff(oldChildren[i], child, element, Component);
    }
  }

  let childLengthDiff = oldChildren.length - newChildren.length;
  let childLength = oldChildren.length - 1;

  while (childLengthDiff > 0) {
    removeNode(Component.node.querySelector(`[${DATA_NODE_ATTRIBUTE}="${oldChildren[childLength].props[DATA_NODE_ATTRIBUTE]}"]`));
    childLength -= 1;
    childLengthDiff -= 1;
  }
};
