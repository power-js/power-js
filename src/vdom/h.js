import { VNode } from './vnode';

/**
 * stack // TODO better description
 * @type {Array}
 */
const stack = [];

/**
 * Returns a new virtual node
 * @param  {String} tagName String containing the elements tag name (i.e. 'div', 'span')
 * @param  {Object} props   Object containing any attributes defined on the element
 * @param  {Array}  args    Array of child nodes
 * @return {Object}         A new virtual node
 */
export function h(tagName, props) {
  const children = [];

  for (let i = arguments.length; i-- > 2; ) {
    stack[stack.length] = arguments[i];
  }

  while (stack.length) {
    let child = stack.pop();

    if (child.pop) {
      for (let i = child.length; i--; ) {
        stack[stack.length] = child[i];
      }
    } else {
      if (typeof child === 'boolean') {
        child = null;
      }

      if (typeof child === 'number') {
        child = String(child);
      }

      if (typeof child !== 'function') {
        if (child === null) {
          child = '';
        }
      }

      children[children.length] = child;
    }
  }

  return new VNode(tagName, props, children);
}
