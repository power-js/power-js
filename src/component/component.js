import { extend } from '../utils/objects/extend';
import { isEqual, isFunction, isObject } from '../utils/is';
import { createElement } from '../dom/createElement';
import { diff } from '../diff/diff';
import { proxy } from './proxy';

/**
 * Power Component
 * @public
 * @class   Component
 */
export class Component {
  /**
   * constructor of class Component
   * @public
   * @param {Object} props
   */
  constructor(props) {
    const construct = this.constructor;

    // Getting called before constructor
    if (this.componentWillInitialize) {
      this.componentWillInitialize(this);
    }

    // mark class as power comp
    this._power = true;

    // the component gets the name of the class name
    this.name = construct.name;

    // default props
    this.props = isFunction(this.getDefaultProps) ? this.getDefaultProps() : isObject(construct.defaultProps) ? construct.defaultProps : {};

    // merge defaultProps with inline props
    if (props) {
      this.props = extend({}, this.props, props);
    }

    // intial state
    this.state = isFunction(this.getInitialState) ? this.getInitialState() : isObject(construct.initialState) ? construct.initialState : {};

    // Getting called after constructor
    if (this.componentDidInitialize) {
      this.componentDidInitialize(this);
    }
  }

  /**
   * creates the elements of a component
   * @public
   * @return {Node}
   */
  create() {
    // convert props into proxy object
    this.props = proxy(this, this.props);

    // get the vnode construct
    this.componentVDom = this.render();

    // create a ref to the element
    this.node = createElement(this.componentVDom, this);

    // return the element
    return this.node;
  }

  shouldComponentUpdate() {
    return true;
  }

  /**
   * Modifies the component state and triggers a re-render of the component
   * and its children with the updated state
   * @public
   * @param {*}        state          The new state Object or the state handler function
   * @param {Function} updateCallback The callback function to invoke after state has been updated
   */
  setState(state, updateCallback) {
    // keep a ref to prevState
    const prevState = this.state;

    if (isEqual(state, prevState)) {
      return;
    }

    const props = this.props;

    let newState = state;
    // if newState is a function
    if (isFunction(newState)) {
      // pass current currentState
      newState = newState.call(this, prevState, props);
    }
    // merge the new state with the existing
    newState = extend({}, prevState, newState);

    // if false, drop any state changes
    if (!this.shouldComponentUpdate(props, newState)) {
      return false;
    }

    // apply state changes
    this.state = newState;

    // update the component
    this.rerender();

    if (isFunction(updateCallback)) {
      updateCallback.call(this);
    }
  }

  /**
   * Forces re-renders the component and its children
   * @public
   */
  forceUpdate(callback) {
    this.rerender();

    if (isFunction(callback)) {
      callback.call(this);
    }
  }

  /**
   * Rerenders the component
   * @public
   */
  rerender() {
    if (this.componentWillUpdate) {
      this.componentWillUpdate(this);
    }

    const updatedComponentVDom = this.render();

    this.patch(this.componentVDom, updatedComponentVDom);

    // assign the new component vdom as the current one
    this.componentVDom = updatedComponentVDom;

    if (this.componentDidUpdate) {
      this.componentDidUpdate(this);
    }
  }

  /**
   * patch
   * @public
   */
  patch(oldVNode, newVNode) {
    diff(oldVNode, newVNode, this);
  }

  /**
   * remove component and its childs
   * @public
   */
  destroy() {
    if (this.componentWillUnmount) {
      this.componentWillUnmount(this);
    }

    this.node.parentNode.removeChild(this.node);

    this.node = null;

    if (this.componentDidUnmount) {
      this.componentDidUnmount(this);
    }
  }
}
