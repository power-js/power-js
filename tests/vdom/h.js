import { h } from '../../src/vdom/h';
import { Component } from '../../src/component/component';

class MyComponent extends Component {
  render() {
    return <div>{this.props.children}</div>;
  }
}

describe('vdom', () => {
  describe('#h', () => {
    it('should return a VNode with the expected attributes', () => {
      const component = <MyComponent />;

      expect(component).toHaveProperty('tagName');
      expect(component).toHaveProperty('children');
      expect(component).toHaveProperty('props');
    });

    it('should contain children of mixed types', () => {
      const component = (
        <MyComponent id="test">
          <MyComponent>
            <div>Hello</div>
            <div>{[<span>World</span>]}</div>
          </MyComponent>
          {true}
        </MyComponent>
      );

      expect(Array.isArray(component.children)).toBe(true);
      expect(component.children[0].constructor.name).toEqual('VNode');
      expect(component.props.id).toEqual('test');
      expect(typeof component.children[1] === 'string').toBe(true);
      expect(component).toMatchSnapshot();
    });

    it('should handle rendering of invalid children (boolean)', () => {
      const component = (
        <MyComponent>
          <MyComponent>Hello</MyComponent>
          {true}
        </MyComponent>
      );

      expect(Array.isArray(component.children)).toBe(true);
      expect(component.children[1]).toEqual('');
      expect(component).toMatchSnapshot();
    });

    it('should handle rendering of invalid children (number)', () => {
      const component = (
        <MyComponent>
          <MyComponent>Hello</MyComponent>
          {1}
        </MyComponent>
      );

      expect(Array.isArray(component.children)).toBe(true);
      expect(component.children[1]).toEqual('1');
      expect(component).toMatchSnapshot();
    });
  });
});
