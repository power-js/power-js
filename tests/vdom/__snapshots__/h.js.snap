// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`vdom #h should contain children of mixed types 1`] = `
VNode {
  "children": Array [
    VNode {
      "children": Array [
        VNode {
          "children": Array [
            "Hello",
          ],
          "props": Object {
            "power-id": 2,
          },
          "tagName": "div",
        },
        VNode {
          "children": Array [
            VNode {
              "children": Array [
                "World",
              ],
              "props": Object {
                "power-id": 3,
              },
              "tagName": "span",
            },
          ],
          "props": Object {
            "power-id": 4,
          },
          "tagName": "div",
        },
      ],
      "props": Object {
        "power-id": 5,
      },
      "tagName": [Function],
    },
    "",
  ],
  "props": Object {
    "id": "test",
    "power-id": 6,
  },
  "tagName": [Function],
}
`;

exports[`vdom #h should handle rendering of invalid children (boolean) 1`] = `
VNode {
  "children": Array [
    VNode {
      "children": Array [
        "Hello",
      ],
      "props": Object {
        "power-id": 7,
      },
      "tagName": [Function],
    },
    "",
  ],
  "props": Object {
    "power-id": 8,
  },
  "tagName": [Function],
}
`;

exports[`vdom #h should handle rendering of invalid children (number) 1`] = `
VNode {
  "children": Array [
    VNode {
      "children": Array [
        "Hello",
      ],
      "props": Object {
        "power-id": 9,
      },
      "tagName": [Function],
    },
    "1",
  ],
  "props": Object {
    "power-id": 10,
  },
  "tagName": [Function],
}
`;
